pkg/controller/controllerutil/controllerutil.go

```diff
@@ -42,7 +64,37 @@ func SetControllerReference(owner, object v1.Object, scheme *runtime.Scheme) err
 	// Create a new ref
 	ref := *v1.NewControllerRef(owner, schema.GroupVersionKind{Group: gvk.Group, Version: gvk.Version, Kind: gvk.Kind})
 
-	// Add it to the child
-	object.SetOwnerReferences(append(object.GetOwnerReferences(), ref))
+	existingRefs := object.GetOwnerReferences()
+	fi := -1
+	for i, r := range existingRefs {
+		if sameOwnerReference(ref, r) {
+			fi = i
+		} else if r.Controller != nil && *r.Controller {
+			return NewAlreadyOwnedError(object, r)
+		}
+	}
+	if fi == -1 {
+		existingRefs = append(existingRefs, ref)
+	} else {
+		existingRefs[fi] = ref
+	}
+
+	// Update owner references
+	object.SetOwnerReferences(existingRefs)
 	return nil
 }
+
+// Returns true if a and b point to the same object
+func sameOwnerReference(a, b v1.OwnerReference) bool {
```

Comment:
Since its comparing the GVKs, better name should be `gvkEquals(a,b)`. Current name reads as `a`, `b` has same owner ?

I wonder if something like this exists the apimachinery already ?